src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Formatting/Engine/AbstractFormatEngine.cs

```diff
@@ -45,21 +46,40 @@ internal abstract partial class AbstractFormatEngine
     internal readonly SyntaxFormattingOptions Options;
     internal readonly TreeData TreeData;
 
+    /// <summary>
+    /// It is very common to be formatting lots of documents at teh same time, with the same set of formatting rules and
+    /// options. To help with that, cache the last set of ChainedFormattingRules that was produced, as it is not a cheap
+    /// type to create.
+    /// </summary>
+    private static readonly object s_gate = new();
+    private static (ImmutableArray<AbstractFormattingRule> formattingRules, SyntaxFormattingOptions options, ChainedFormattingRules chainedRules) s_lastRulesAndOptions;
+
     public AbstractFormatEngine(
         TreeData treeData,
         SyntaxFormattingOptions options,
-        IEnumerable<AbstractFormattingRule> formattingRules,
+        ImmutableArray<AbstractFormattingRule> formattingRules,
         SyntaxToken startToken,
         SyntaxToken endToken)
         : this(
               treeData,
               options,
-              new ChainedFormattingRules(formattingRules, options),
+              GetChainedFormattingRules(formattingRules, options),
               startToken,
               endToken)
     {
     }
 
+    private static ChainedFormattingRules GetChainedFormattingRules(ImmutableArray<AbstractFormattingRule> formattingRules, SyntaxFormattingOptions options)
+    {
+        lock (s_gate)
+        {
+            if (formattingRules != s_lastRulesAndOptions.formattingRules || options != s_lastRulesAndOptions.options)
```

Comment:
>if (formattingRules != s_lastRulesAndOptions.formattingRules || options != s_lastRulesAndOptions.options) [](http://example.com/codeflow?start=12&length=105)

Would you still need the lock if s_lastRulesAndOptions was a Tuple instead of a ValueTuple?

Reply:
Nope. Will change